src/emmc.c

```diff
@@ -30,42 +36,54 @@ static int r_emmc_read_extcsd(int fd, guint8 extcsd[512])
 
 gboolean r_emmc_read_bootpart(const gchar *device, gint *bootpart_active, GError **error)
 {
-	gint ret;
-	gboolean res = FALSE;
 	guint8 extcsd[512];
-	int fd;
+	g_auto(filedesc) fd = -1;
 	/* count from 1 */
 	gint active_partition = -1;
 
 	fd = g_open(device, O_RDONLY);
 	if (fd == -1) {
 		int err = errno;
-		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+		g_set_error(error, G_FILE_ERROR, g_file_error_from_errno(err),
 				"opening eMMC device failed: %s", g_strerror(err));
-		goto out;
+		return FALSE;
 	}
 
-	ret = r_emmc_read_extcsd(fd, extcsd);
-	if (ret) {
-		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
-				"Could not read from extcsd register %d in %s\n",
+	if (r_emmc_read_extcsd(fd, extcsd)) {
+		g_set_error(error, R_EMMC_ERROR, R_EMMC_ERROR_IOCTL,
+				"Could not read from extcsd register %d in %s",
 				EXT_CSD_PART_CONFIG, device);
-		goto out;
+		return FALSE;
 	}
 
 	/* retrieve active partition from BOOT_PART register */
 	active_partition = (extcsd[EXT_CSD_PART_CONFIG] & 0x38) >> 3;
 
-	/* return the partitions counted from 0 */
-	*bootpart_active = active_partition - 1;
-
-	res = TRUE;
-
-out:
-	if (fd >= 0)
-		g_close(fd, NULL);
+	switch (active_partition) {
+		case 0x0: /* not boot enabled */
+			*bootpart_active = -1;
+			return TRUE;
+		case 0x1: /* boot0 / boot 1 */
+		/* fallthgrough */
+		case 0x2:
+			/* return the partitions counted from 0 */
+			*bootpart_active = active_partition - 1;
+			return TRUE;
+		case 0x7: /* user data area */
+			g_set_error(error,
+					R_EMMC_ERROR,
+					R_EMMC_ERROR_BOOTPART_UDA,
+					"Active eMMC partition is UDA when boot0/boot1 was expected.");
+			return FALSE;
+		default:
+			g_set_error(error,
+					R_EMMC_ERROR,
+					R_EMMC_ERROR_BOOTPART_INVALID,
+					"Invalid (Reserved) eMMC boot part number.");
+			return FALSE;
+	}
 
-	return res;
+	return TRUE; /* unreachable */
```

Comment:
This isn't the successful case.

```suggestion
	g_return_val_if_reached(FALSE);
```